### PR TITLE
fix(bulk-create): `ON CONFLICT` with unique index

### PR DESCRIPTION
**Pull request by papb**
Supersedes #12516

Also addresses https://github.com/sequelize/sequelize/pull/11984/files#r414863957 via refactor

### CODE REVIEW

File: lib/model.js [hunk 1]
```diff
@@ -2683,11 +2683,22 @@ class Model {
...
           options.updateOnDuplicate = options.updateOnDuplicate.map(attr => model.rawAttributes[attr].field || attr);
-          // Get primary keys for postgres to enable updateOnDuplicate
-          options.upsertKeys = _.chain(model.primaryKeys).values().map('field').value();
-          if (Object.keys(model.uniqueKeys).length > 0) {
-            options.upsertKeys = _.chain(model.uniqueKeys).values().filter(c => c.fields.length >= 1).map(c => c.fields).reduce(c => c[0]).value();
+
+          const upsertKeys = [];
+
+          for (const i of model._indexes) {
+            if (i.unique) {
```
Comment (AllAwesome497) [inline lib/model.js:2690]: probably worth making sure that the index isn't a partial index - if its a partial index then it won't apply unless that where clause is specified, which currently we don't do.

File: lib/model.js [hunk 2]
```diff
@@ -2683,11 +2683,22 @@ class Model {
...
+            }
           }
+
+          const firstUniqueKey = Object.values(model.uniqueKeys).find(c => c.fields.length >= 1);
+
+          if (firstUniqueKey && firstUniqueKey.fields) {
+            upsertKeys.push(...firstUniqueKey.fields);
+          }
+
+          options.upsertKeys = upsertKeys.length > 0 ? upsertKeys : Object.values(model.primaryKeys).map(x => x.field);
```
Comment (AllAwesome497) [inline lib/model.js:2701]: small nit:
```suggestion
          options.upsertKeys = upsertKeys.length > 0
            ? upsertKeys
            : Object.values(model.primaryKeys).map(x => x.field);
```

File: lib/model.js (hunk 1 above)
Comment (papb) [reply lib/model.js]: Done at 9bc372a

File: lib/model.js [hunk 3]
```diff
@@ -2683,11 +2683,24 @@ class Model {
...
+
+          const upsertKeys = [];
+
+          for (const i of model._indexes) {
+            if (i.unique && !i.where) { // Don't infer partial indexes
+              upsertKeys.push(...i.fields);
+            }
           }
+
+          const firstUniqueKey = Object.values(model.uniqueKeys).find(c => c.fields.length >= 1);
```
Comment (AllAwesome497) [inline lib/model.js:2695]: personally I prefer `c.fields.length > 0` but not a big deal.

**Review by AllAwesome497 — APPROVED** [review]
Lgtm :+1:

**Pull request merged**